Set explicit z-index on interface body to ensure it’s pinned under interface header

getdave opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description

On smaller viewports on Safari iOS the interface header can (under limited circumstances) suffer a render glitch whereby the content below "bleeds" into it.

See screenshots for Before/After.

This PR fixes this by setting an explicit `z-index` for the interface skeleton ~"body"~ "content" and ensuring this is lower than that of the interface skeleton "header".

**Update**: we had to revert applying to the "body" because that caused the "actions" panel to sit under the header when publishing a post. Applying to "content" achieves the same fix but without the nasty knock on effects.

Fixes https://github.com/WordPress/gutenberg/issues/32631

## How has this been tested?

Either on a real iOS device on the iOS simular spin up Safari and load WP with this branch built and enabled.

### Before - verify the bug
* Using Twenty Fourteen as a theme
* Go to Appearance > Widgets
* Make sure to add some blocks to be able to scroll the page
* Scroll slowly
* Notice the black top bar sometimes appears

### After - verify the fix
* Using Twenty Fourteen as a theme
* Go to Appearance > Widgets
* Make sure to add some blocks to be able to scroll the page
* Scroll slowly
* Notice the top bar retains the "white" background and no black bleeds through.

Also verify this change does not break other screens such as Post Editor or Site Editor.

## Screenshots <!-- if applicable -->

### Before
https://user-images.githubusercontent.com/444434/122210810-f1644600-ce9d-11eb-9b22-1d0498d964e5.mp4

### After
https://user-images.githubusercontent.com/444434/122210806-f0cbaf80-ce9d-11eb-8b90-2f350190ed07.mp4





## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

getdave committed:
Set explicit z-index on interface body to ensure it’s pinned under the interface header

jasmussen commented:
Thanks for thinking about these issues. We've had a few Safari related issues lately, so I'd appreciate if @gwwar could also take this for a quick spin or just be aware of it.

But from I can tell, this change is fine. It's a global change that affects all editors, so I focused mostly to see on whether there were any regressions in post or site editors, there don't appear to be. Here's Chrome and Safari:

![saf testing](https://user-images.githubusercontent.com/1204802/122215198-3b9ff400-ceab-11eb-81fc-6721bee60e55.gif)
![testing](https://user-images.githubusercontent.com/1204802/122215244-478bb600-ceab-11eb-8d64-f52fa13866cf.gif)

In general I tested:

- The various configurations of the skeleton, with and without top toolbar, with and without fullscreen
- Those changes but with the various breakpoints we have
- Scrolling block toolbars and items that might overlap or overlay the top toolbar

As noted, I don't see any issues with this, so I think we can 👍 👍 this one after Kerry looks. Thanks Dave!

gwwar commented:
I'll try to 👀 today. If other folks would like to help I think we'll want to verify behavior on a native iOS device, a native android device and our normal browser desktop variants. 

For context, I noticed that there were a number of browser issues related to z-index and --webkit-overflow-scrolling, for example https://bugs.chromium.org/p/chromium/issues/detail?id=128325 or in webkit: https://bugs.webkit.org/buglist.cgi?quicksearch=touch%20z-index

jasmussen commented:
> If other folks would like to help I think we'll want to verify behavior on a native iOS device, a native android device and our normal browser desktop variants.

I wanted to suggest that if you are on a Mac you can install Xcode for free from the app store and use the built-in simulator, which should be as good as a physica device:

<img width="463" alt="Screenshot 2021-06-16 at 18 02 00" src="https://user-images.githubusercontent.com/1204802/122254112-17eda580-cecd-11eb-9460-7f4fde98dd6b.png">

<img width="678" alt="Screenshot 2021-06-16 at 18 02 21" src="https://user-images.githubusercontent.com/1204802/122254122-191ed280-cecd-11eb-983d-377bb3c025e0.png">

The benefit is that it's networked with your Mac so you can just type in `localhost` in Safari and it will bridge to your local dev environment. You can even inspect the contents of Mobile Safari using the Safari dev tools:

<img width="1179" alt="Screenshot 2021-06-16 at 18 03 55" src="https://user-images.githubusercontent.com/1204802/122254255-3d7aaf00-cecd-11eb-823a-0406c05c2815.png">

The downside is that Xcode takes half a day to download and install. But it's worth it.

_I assume you all know this already, but figured I'd document it for any others viewing the PR._

gwwar commented:
Visiting http://gutenberg.run/32732 should do the trick too, if folks are using a physical phone.

getdave commented:
Just to say both videos in my PR desc are using the iOS simulator technique.

gwwar commented:
On a physical pixel 4: This might be an unrelated bug, but I can easily get a paragraph stuck, where an <kbd>enter</kbd> keypress doesn't create a new paragraph. An easy way to do this is to select a paragraph after a scroll then try typing <kbd>enter</kbd>

https://user-images.githubusercontent.com/1270189/122289393-861f7180-cea7-11eb-87fa-8553b48aa570.mp4

gwwar commented:
likely another unrelated one: On ios simulator, adding an image to the widgets screen makes the control width incorrect:


https://user-images.githubusercontent.com/1270189/122291522-c41d9500-cea9-11eb-9194-ebbb439546f4.mp4

gwwar reviewed (approved):
Nice work! Going to give tentative approval, much appreciated if we could clear up the E2E tests

Verified that this fixes the black scrollbar in iOS:

Before:

https://user-images.githubusercontent.com/1270189/122293173-9afe0400-ceab-11eb-8096-4aa7ba95bdde.mp4

After:

https://user-images.githubusercontent.com/1270189/122293225-a81af300-ceab-11eb-8764-cef30ee30f8a.mp4

getdave commented:
I have a real problem with the e2e tests here. Basically the e2e helper that clicks the publish button and waits for the snackbar to popup and say "Published" is failing.

The reason for this (AFAIK) is that the "Publish" action simply isn't being triggered. I did think it was the snackbar not appearing due to the new `z-index` but if you run the tests locally in interactive mode you can see the Publish action never happens.

If you try and mimic manually you will see that clicking `Publish` doesn't actually "publish" anymore. Rather it just toggles the sidebar. However, switch to master and rebuild and you will see everything works again. 

I don't know _what_ is going on here.

jasmussen commented:
Have you tried interactive tests? I can't recall the puppetteer attributes, but you can run them interactively so they show a small chromium window running each test. That might visualize the problem.

gwwar commented:
Yup, 💯 for interactive tests as @jasmussen suggested.

To run this locally, make sure wp-env is started (visit localhost:8889 to make sure things look okay and gutenberg is activated). 

Then type:
`npm run test-e2e -- --puppeteer-interactive specs/editor/various/preview.test.js`. (See other options in https://developer.wordpress.org/block-editor/contributors/code/testing-overview/#end-to-end-testing)

If the suite is long it can help to add a temporary `it.only` to the affected test. @getdave let me know if you get stuck and I can help 🔍

getdave commented:
Thanks for the advice both. In fact that is exactly what I've been doing and indeed it was what first allowed me to narrow down the issue.

I've now pushed up a "fix" which gets (or at least should get!) the tests to pass. The question is though why do we _need_ this fix just because I added a `z-index`. To me it's currently making absolutely no sense.

Prior to the fix here was the flow for the publish button:

* The value of the `isToggle` prop for the `<PostPublishButton>` is set here
https://github.com/WordPress/gutenberg/blob/d91786ae138b6c1466498f2734ac0583ce5b0d78/packages/edit-post/src/components/header/post-publish-button-or-toggle.js#L77
* The value of the `isToggle` is set based on this logic:
https://github.com/WordPress/gutenberg/blob/d91786ae138b6c1466498f2734ac0583ce5b0d78/packages/edit-post/src/components/header/post-publish-button-or-toggle.js#L58-L70
* The value of `isToggle` prop of `<PostPublishButton>` determines which button is shown
https://github.com/WordPress/gutenberg/blob/d91786ae138b6c1466498f2734ac0583ce5b0d78/packages/editor/src/components/post-publish-button/index.js#L176
* [`buttonProps`](https://github.com/WordPress/gutenberg/blob/d91786ae138b6c1466498f2734ac0583ce5b0d78/packages/editor/src/components/post-publish-button/index.js#L149-L155) turns the button in one which actually Publishes.
* [`toggleProps`](https://github.com/WordPress/gutenberg/blob/d91786ae138b6c1466498f2734ac0583ce5b0d78/packages/editor/src/components/post-publish-button/index.js#L157-L164) turns the button in one which toggles the pre-publoish sidebar.

In this PR prior to my fix to `<PostPublishButtonOrToggle>` the `isToggle` value was always ending up as `true` meaning the Publish button simply toggled the sidebar open/closed. I think this is because

* `isPublishSidebarEnabled` is a editor preference setting that indicates whether or not to _show_ the panel  _at all_ (rather than whether or not it's actually _open_).
* `isSmallerThanMediumViewport` is a boolean so will _always_ be true if the viewport is smaller that `780px`.

With my fix in place which has an _additional_ check to see whether the panel is already _open_ we give the logic a chance of allowing the button to be a true "Publish" button.

What I don't understand is why it's necessary. I can't see this happen in `trunk`. It seems confined to this branch despite me having rebased it several times against the latest `trunk`.

gwwar commented:
My initial thought is that some event propagation has changed (for example a component that should be listening to the click doesn't get it). I'll try to help 🔍 later today

getdave commented:
I have a suspicion that we're expecting `isPublishSidebarEnabled` to mean "is the publish sidebar currently visible".

draganescu commented:
We shouldn't need to add the second fix. My guess is similar to @gwwar 's that the `z-index` makes it that something does not receive the click anymore, and something else does. It's pretty unlikely that this tiny change uncovered a logic error in the header component.

gwwar commented:
Stepping through, the difference is that when the publish panel is open and we click the button again, `isPublished` should be true.

| publish click with z-index | publish click no z-index | data store |
|-----|-----|-----|
| <img width="837" alt="publishz" src="https://user-images.githubusercontent.com/1270189/122591165-6dce6480-d017-11eb-8d8c-cef7ff0c90ea.png"> | <img width="840" alt="publish no z" src="https://user-images.githubusercontent.com/1270189/122591181-73c44580-d017-11eb-948a-93e7bde6f798.png"> | <img width="1162" alt="Screen Shot 2021-06-18 at 9 26 46 AM" src="https://user-images.githubusercontent.com/1270189/122591217-7de64400-d017-11eb-931a-af4c49ef8c4e.png"> |

@getdave I'd recommend looking at the core/editor store dispatches next. Let me know if you're not able to track it down. For debugging, what works for me is observing the flow of state updates for the related reducer in trunk, and seeing what dispatched the event. After doing so I'd compare with the branch and see if any dispatches are missing or incorrect.

gwwar reviewed (changes requested):
E2E tests had a great catch here ✨ . We'll want to fix being able to publish a post, and making sure that the data stores are in the right state. Please ping me for another look when this is ready, or if you'd like help 🔍 further.

getdave commented:
> Stepping through, the difference is that when the publish panel is open and we click the button again, isPublished should be true.

If this is the case then it makes no sense. The `isPublished` status should not be tied to whether the panel is open or not. Moreover, we should not rely on clicking the button to change the status of `isPublished` in time for the action of the button to be updated to be "Publish" rather than "Toggle". 

I think perhaps we have a race condition. Diving deeper...

getdave commented:
I have the answer...incoming explanation 🥳

getdave commented:
| publish click with z-index | publish click no z-index | 
|-----|-----|
| https://user-images.githubusercontent.com/444434/122763767-a0d25b80-d296-11eb-8833-30b32fb7531b.mp4 | https://user-images.githubusercontent.com/444434/122763785-a4fe7900-d296-11eb-844e-7c115572a26a.mp4 | 


Something
https://user-images.githubusercontent.com/444434/122763789-a5970f80-d296-11eb-9808-c253795c22b4.mp4

getdave commented:
## Direct Comparison

Essentially the issue is that we're not comparing apples with apples. Or to put it another way we aren't looking at the same Publish button when we're comparing `z-index` with no `z-index`. They are completely different buttons.

What's happening is the `z-index` is not allowing the correct publish button (ie: the one that you see on `trunk`) to show up as it's stacked underneath the body:

https://user-images.githubusercontent.com/444434/122763977-de36e900-d296-11eb-8fb3-5f22e6563fea.mp4


### With z-index
https://user-images.githubusercontent.com/444434/122763990-e2fb9d00-d296-11eb-96ea-2c82a7b1a081.mp4

### Without z-index
https://user-images.githubusercontent.com/444434/122763997-e42cca00-d296-11eb-9abc-83c264a8bb2c.mp4

getdave committed:
Add z-index to skelton content to avoid publish button bug

Adding z-index to body causes Publish button in actions sidebar to appear underneath the layout header causing the “wrong” publish button to be shown. Applying z-index to content avoids this whilst achieving the same outcome in terms of avoiding the bleed of backgrounfd on mobile safari when scrolling the Widgets editor.

getdave commented:
@gwwar I've narrowed down the issue and implemented a fix. It was a weird one but so obvious once you see it!

Are you still happy with your 👍 ?

The e2e failures are being caused by https://github.com/WordPress/gutenberg/pull/32857

gwwar commented:
@getdave which fix are you referring to, did you push to the branch?

getdave commented:
> @getdave which fix are you referring to, did you push to the branch?

@gwwar yes. It is very minor but zindex now applied to skeleton content not body.

Looks like I forgot to rename the CSS var but I can do that tomorrow.

Also see explanation in comments above.

gwwar commented on packages/interface/src/components/interface-skeleton/style.scss:
```diff
@@ -82,9 +82,13 @@ html.interface-interface-skeleton__html-container {
+	// On Safari iOS on smaller viewports lack of a z-index causes the background
+	// to "bleed" through the header.
+	// See https://github.com/WordPress/gutenberg/issues/32631
+	z-index: z-index(".interface-interface-skeleton__body");
```
Oh, gotcha the diff was hiding that this is being applied to `.interface-interface-skeleton__content`.

getdave replied:
Yeh I'm sorry. It was unclear because I forgot about the naming of this CSS var `.interface-interface-skeleton__body`. It should be `.interface-interface-skeleton__content`. Looks like @noisysocks sorted it after he merged early.

gwwar reviewed (approved):
I retested this in ios simulator, and verified that this fixes the background behavior. I also verified in Chrome/Safari that we can create a new post and publish.

Much appreciated if we can update the z-index name before 🚢 but no need for a re-review from me when that happens 👍
